### ISSUE TITLE
Held Holdings - Bound Withs - holding records when all parents have arrived

### ISSUE THREAD

**Issue by patrickzurek** [eXtensibleCatalog/xcmetadataservicestoolkit#555]
JIRA issue created by: rcook
Originally opened: 2011-08-31 01:44 PM

Issue body:
 (nt)

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-08-31 01:45 PM

Comment body:

John searched through held_holdings database table looking for which manifestation_id a held_holding_id was waiting on and the one he checked, the manifestation had arrived.  Seems to be a bug.

From: Cook, Randall 
Sent: Tuesday, August 30, 2011 5:17 PM
To: Bowen, Jennifer; Brand, John
Cc: Lindahl, David; Anderson, Benjamin D
Subject: bound with that has 2 parents

I labored through finding the parents for this bound with.  As far as I can tell, this holdings has two parent bibs and both were processed so I canâ€™t figure out why it is held.  If nothing jumps out at anyone, I will open a FB issue.  Maybe the Transform service  is not fixed to appropriately deal and match for bound withs.

 
Held holding 4633143
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=11945447&query=&searchXML=false&selectedFacetNames=|successor&selectedFacetValues=|24094876&rowStart=0&startPageNumber=1&currentPageNumber=1 

<marc:record xmlns:marc="http://www.loc.gov/MARC21/slim" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/MARC21/slim http://128.151.244.137:8080/OAIToolkit/schema/MARC21slim_custom.xsd">
  <marc:leader>00361cx  a22000974  4500</marc:leader>
  <marc:controlfield tag="001">4633143</marc:controlfield>
  <marc:controlfield tag="003">NRU</marc:controlfield>
  <marc:controlfield tag="004">4361545</marc:controlfield>
  <marc:controlfield tag="005">20100128103742.0</marc:controlfield>
  <marc:controlfield tag="008">1001284u    8   1001uu   0901128</marc:controlfield>
  <marc:datafield ind1="1" ind2=" " tag="014">
    <marc:subfield code="a">4361595</marc:subfield>
  </marc:datafield>
  <marc:datafield ind1="0" ind2=" " tag="852">
    <marc:subfield code="b">SibRare</marc:subfield>
    <marc:subfield code="h">ML410.M53</marc:subfield>
    <marc:subfield code="i">B46</marc:subfield>
    <marc:subfield code="z">Bound together: Benedict, J. Sketch of the life and works of the late Felix Mendelssohn Bartholdy and Chorely, H.F. The pianoforte.</marc:subfield>
    <marc:subfield code="x">Accession no.: 446536.</marc:subfield>
  </marc:datafield>
</marc:record>


 BIB 4361595 was processed
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=11945517&query=&searchXML=false&selectedFacetNames=|successor&selectedFacetValues=|24095048&rowStart=0&startPageNumber=1&currentPageNumber=1 

BIB 4361545 was processed

http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=11945446&query=&searchXML=false&selectedFacetNames=|successor&selectedFacetValues=|24094873&rowStart=0&startPageNumber=1&currentPageNumber=1

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-10-04 03:53 PM

Comment body:

Need to do more investigation to accurately estimate this one.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-10-26 04:59 PM

Comment body:

Appears to work right only if the bibs arrive 1st.  This I have found through unit testing locally.  Put in a potential fix, need to test it on a grander scale.  All the pieces seemed to be in place, but they were not utilized correctly.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-10-26 05:07 PM

Comment body:

Also note, fixed another problem (last week) where the long value was not getting parsed correctly from the string value.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-11-17 04:36 PM

Comment body:

John, point me to the newest full run and I can research and find more.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-11-28 02:37 PM

Comment body:

Assigning to RC to do some analysis.  Check out http://128.174.70.247:8080/MetadataServicesToolkit/st/viewLogin.action, there are 10,884 held.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-11-28 07:10 PM

Comment body:

I think it looks like parents 3421995, 3421996, 3421998, and 3421999 have arrived.

<marc:record xmlns:marc="http://www.loc.gov/MARC21/slim" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/MARC21/slim http://128.151.244.137:8080/OAIToolkit/schema/MARC21slim_custom.xsd">
  <marc:leader>00311cx  a22001214  4500</marc:leader>
  <marc:controlfield tag="001">3611684</marc:controlfield>
  <marc:controlfield tag="003">NRU</marc:controlfield>
  <marc:controlfield tag="004">3421995</marc:controlfield>
  <marc:controlfield tag="005">20070112133404.0</marc:controlfield>
  <marc:controlfield tag="008">0701124u    8   1001ub   0901128</marc:controlfield>
  <marc:datafield ind1="1" ind2=" " tag="014">
    <marc:subfield code="a">3421996</marc:subfield>
  </marc:datafield>
  <marc:datafield ind1="1" ind2=" " tag="014">
    <marc:subfield code="a">3421998</marc:subfield>
  </marc:datafield>
  <marc:datafield ind1="1" ind2=" " tag="014">
    <marc:subfield code="a">3421999</marc:subfield>
  </marc:datafield>
  <marc:datafield ind1="8" ind2=" " tag="852">
    <marc:subfield code="b">SibMicAcq</marc:subfield>
    <marc:subfield code="h">Film 1109</marc:subfield>
    <marc:subfield code="z">Film contains 4 essays by Le Sueur. Sibley cannot reproduce.</marc:subfield>
  </marc:datafield>
</marc:record>

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-11-28 07:15 PM

Comment body:

Here is another, I think parents 3422245 and 3422246 have both been processed and produced output, but the holdings is held.  I will discontinue testing at this point.



<marc:record xmlns:marc="http://www.loc.gov/MARC21/slim" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/MARC21/slim http://128.151.244.137:8080/OAIToolkit/schema/MARC21slim_custom.xsd">
  <marc:leader>00261cx  a22000974  4500</marc:leader>
  <marc:controlfield tag="001">3611973</marc:controlfield>
  <marc:controlfield tag="003">NRU</marc:controlfield>
  <marc:controlfield tag="004">3422245</marc:controlfield>
  <marc:controlfield tag="005">20070115152318.0</marc:controlfield>
  <marc:controlfield tag="008">0701154u    8   1001ub   0901128</marc:controlfield>
  <marc:datafield ind1="1" ind2=" " tag="014">
    <marc:subfield code="a">3422246</marc:subfield>
  </marc:datafield>
  <marc:datafield ind1="8" ind2=" " tag="852">
    <marc:subfield code="b">SibMicAcq</marc:subfield>
    <marc:subfield code="h">Film 323</marc:subfield>
    <marc:subfield code="z">Film contains Libro 1 and Libro 2. Sibley cannot reproduce.</marc:subfield>
  </marc:datafield>
</marc:record>




Randy,

Can you see if these already-processed parent bibs (3422245, 3422246) are listed in the bibsYet2ArriveLongId mysql table?  (I don't have an account on server 128.174.70.247).

echo "select * from bibsYet2ArriveLongId" | mysql -uroot -proot xc_marctoxctransformation

[The above lists all bib001's of for the bibs marctoxctransformation is still waiting on.  The following will only list the parent bibs if they exist:]

echo "select * from bibsYet2ArriveLongId where bib_001 in (3422245, 3422246)" | mysql -uroot -proot xc_marctoxctransformation

I have a sneaky suspicion they are not listed (because they were actually processed correctly).  I think this is a SOLR indexing bug (the SOLR record needs to be changed from held to active).

Because it seems to me all of the output records were created properly:

1) Here's the holdings marc record (marcnormalization) with 001 of 3611973:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=11279071

<marc:record xmlns:marc="http://www.loc.gov/MARC21/slim" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/MARC21/slim http://128.151.244.137:8080/OAIToolkit/schema/MARC21slim_custom.xsd">
  <marc:leader>00261cx  a22000974  4500</marc:leader>
  <marc:controlfield tag="001">3611973</marc:controlfield>
  <marc:controlfield tag="003">NRU</marc:controlfield>
  <marc:controlfield tag="004">3422245</marc:controlfield>
  <marc:controlfield tag="005">20070115152318.0</marc:controlfield>
  <marc:controlfield tag="008">0701154u    8   1001ub   0901128</marc:controlfield>
  <marc:datafield ind1="1" ind2=" " tag="014">
    <marc:subfield code="a">3422246</marc:subfield>
  </marc:datafield>
  <marc:datafield ind1="8" ind2=" " tag="852">
    <marc:subfield code="b">SibMicAcq</marc:subfield>
    <marc:subfield code="h">Film 323</marc:subfield>
    <marc:subfield code="z">Film contains Libro 1 and Libro 2. Sibley cannot reproduce.</marc:subfield>
  </marc:datafield>
</marc:record>

2) ...and its successor xc holdings record (marctoxctransformation):
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701319

<xc:frbr xmlns:xc="http://www.extensiblecatalog.info/Elements" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:rdarole="http://rdvocab.info/roles" xmlns:rdvocab="http://rdvocab.info/Elements" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <xc:entity id="oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701319" type="holdings">
    <xc:location>SibMicAcq</xc:location>
    <xc:callNumber>Film 323</xc:callNumber>
    <xc:recordID type="NRU">3611973</xc:recordID>
    <xc:manifestationHeld>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701315</xc:manifestationHeld>
    <xc:manifestationHeld>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701318</xc:manifestationHeld>
  </xc:entity>
</xc:frbr>


3) Parent#1's manifestation:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701315

<xc:frbr xmlns:xc="http://www.extensiblecatalog.info/Elements" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:rdarole="http://rdvocab.info/roles" xmlns:rdvocab="http://rdvocab.info/Elements" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <xc:entity id="oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701315" type="manifestation">
    <xc:recordID type="OCoLC">2752988</xc:recordID>
    <xc:recordID type="NRU">3422245</xc:recordID>
    <dcterms:title>Liber capelle Franchini Gafori [vol.1].</dcterms:title>
    <dcterms:alternative>1 Sanctus, 23 Magnificats, 4 Magnificat sections, 1 Te Deum, 7 hymns, 104 motets (some arranged in cycles, as substitutes for Mass sections), 2 textless pieces</dcterms:alternative>
    <dcterms:issued>[ca. 1484-1490]</dcterms:issued>
    <dcterms:extent>188 leaves.</dcterms:extent>
    <dcterms:description>Contains parts for 4-5 voices.</dcterms:description>
    <dcterms:description>Copied in Milan, for use by the cathedral choir, under the direction of Franchino Gafori, maestro di cappella from 1484-1522.</dcterms:description>
    <dcterms:description>Original held in Milan at the Archivio della Veneranda Fabbrica del Duomo, Sezione Musicale: Librone 1 (olim 2269).</dcterms:description>
    <dcterms:description>Contains pieces by: Binchois, CompeÌ�re, Dufay, Gaffurio, Martini, Weerbecke, et al et anon.</dcterms:description>
    <dcterms:description>For contents see: Sartori, Claudio. La Cappella Musicale del Duomo di Milano: Catalogo delle musiche dell'archivio (Milan, 1957), p. 43-50.</dcterms:description>
    <xc:typeLeader06>Manuscript notated music</xc:typeLeader06>
    <xc:type007>Notated music</xc:type007>
    <xc:type007>Microform</xc:type007>
    <xc:typeSMD>Microfilm reel</xc:typeSMD>
    <rdvocab:modeOfIssuance>Monograph/Item</rdvocab:modeOfIssuance>
    <xc:expressionManifested>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701317</xc:expressionManifested>
  </xc:entity>
</xc:frbr>

4) Parent#2's manifestation:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701318

<xc:frbr xmlns:xc="http://www.extensiblecatalog.info/Elements" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:rdarole="http://rdvocab.info/roles" xmlns:rdvocab="http://rdvocab.info/Elements" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <xc:entity id="oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701318" type="manifestation">
    <xc:recordID type="OCoLC">2752960</xc:recordID>
    <xc:recordID type="NRU">3422246</xc:recordID>
    <dcterms:title>Liber capelle Franchini Gafori [vol.2].</dcterms:title>
    <dcterms:alternative>11 Masses (1 without Kyrie; 1 without Agnus), 11 G-C-S Masses, 1 Gloria-Credo pair, 4 Sanctus, 1 Te Deum, 12 motets (some intended as substitutes for Mass sections)</dcterms:alternative>
    <dcterms:issued>[ca. 1490-1520]</dcterms:issued>
    <dcterms:extent>211 leaves.</dcterms:extent>
    <dcterms:description>Contains parts for 4-5 voices.</dcterms:description>
    <dcterms:description>Copied in Milan, for use by the cathedral choir, under the direction of Franchino Gafori, maestro di cappella from 1484-1522.</dcterms:description>
    <dcterms:description>Original held in Milan at the Archivio della Veneranda Fabbrica del Duomo, Sezione Musicale: Librone 2 (olim 2268).</dcterms:description>
    <dcterms:description>Contains music by: Brumel, CompeÌ€re, Gafori, Isaac, Martini, Obrecht, Tinctoris, Weerbecke, et anon.</dcterms:description>
    <dcterms:description>For contents see: Sartori, Claudio. La Cappella Musicale del Duomo di Milano: Catalogo delle musiche dell'archivio (Milan, 1957), p. 43-50.</dcterms:description>
    <xc:typeLeader06>Manuscript notated music</xc:typeLeader06>
    <xc:type007>Notated music</xc:type007>
    <xc:type007>Microform</xc:type007>
    <xc:typeSMD>Microfilm reel</xc:typeSMD>
    <rdvocab:modeOfIssuance>Monograph/Item</rdvocab:modeOfIssuance>
    <xc:expressionManifested>oai:mst.rochester.edu:MetadataServicesToolkit/marctoxctransformation/22701321</xc:expressionManifested>
  </xc:entity>
</xc:frbr>

5) ...and all the referenced expressions and works for both these parent manifestions were also created:

Parent#1's expression:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701317

Parent#1's work:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701316

Parent#2's expression:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701321

Parent#2's work:
http://128.174.70.247:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=22701320




--Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-12-07 04:53 PM

Comment body:

Chris, moving this to you as it might make sense for you to finish the "full work needed" to get your held holdings number under control.

Just to be clear about a nuance I think that exists here.

1.  This issue was opened because there were cases when all parent bibs were in the process but some holdings were still held.  These were records that were in the record set to begin with.
2.  The nuance here is that there is a bug where subsequent arrival of a missing bib does not trigger the correct reprocessing of a held record.

These may be the same root cause or may be different.  I just wanted to spell that out.


From: Brand, John 
Sent: Wednesday, December 07, 2011 11:35 AM
To: 'Delis, Christopher'; Cook, Randall; Bowen, Jennifer
Cc: Lindahl, David
Subject: RE: FogBugz (MST-354) MST - Holding records 014 with leading configurable chars

It is 823 assigned to me for the next release.

http://xcfogbugz.lib.rochester.edu/default.asp?823

From: Delis, Christopher [mailto:cedelis@uillinois.edu] 
Sent: Wednesday, December 07, 2011 11:31 AM
To: Brand, John; Cook, Randall; Bowen, Jennifer
Cc: Lindahl, David
Subject: RE: FogBugz (MST-354) MST - Holding records 014 with leading configurable chars

I ran a test to see if MST would be able to process held records if I were to import those missing 67K bibs.  I was able to import almost all of them (except ~200 which were invalid marc), but in the end I ended up with exactly the same number of held records (137,830) as before.  I was expecting this number to drop to near 200.  In my example below, you can see that bib 1287266 now exists in the MST: 

http://xc.ilcso.uiuc.edu:8080/MetadataServicesToolkit/st/viewRecord.action?recordId=51737072&query=1287266&selectedFacetNames=&selectedFacetValues=&rowStart=0&startPageNumber=1Â¤tPageNumber=1

It's possible I'm not understanding something about the process.  Also,
it was mentioned that there is already a known bug regarding held records, but I can't seem to find it (can someone point me to it please?). 

Chris


________________________________________
From: Delis, Christopher
Sent: Tuesday, December 06, 2011 1:53 PM
To: Brand, John; Cook, Randall; Bowen, Jennifer
Cc: Lindahl, David
Subject: RE: FogBugz (MST-354) MST - Holding records 014 with leading configurable chars
More info:

Out of the 128K held records, there are 67,203 missing parent bibs.

Out of the 67,203 missing parent bibs, only 351 bibs were actually found to be in the repository.

This means the vast majority (66,852) somehow never made it into the repo to begin with.  I need to create a new repo with much better "coverage."

Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2011-12-16 08:04 PM

Comment body:

I think there may be an additional bug related to how the bibsYet2ArriveLongIdMap objects are being read from the mysql table (in TransformationService::setup()).  Seems like org_code isn't being selected in TransformationDAO::loadBibMaps().  This might be a cause for some other unnecessary held items.

Here's a diff of what I think needs to be changed:

Index: 

xc/mst/services/transformation/dao/TransformationDAO.java
===================================================================
--- xc/mst/services/transformation/dao/TransformationDAO.java   (revision 2841)
+++ xc/mst/services/transformation/dao/TransformationDAO.java   (working copy)
@@ -215,7 +215,7 @@
         TimingLogger.start("getBibMaps");
         int recordsAtOnce = 250000;
         List<Map<String, Object>> rowList = this.jdbcTemplate.queryForList(
-                "select bib_001, record_id from " + tableName +
+                "select org_code, bib_001, record_id from " + tableName +
                         " limit " + (page * recordsAtOnce) + "," + recordsAtOnce);
         TimingLogger.stop("getBibMaps");
         return rowList;



Today, I verified that the above patch does indeed solve the persistence problem that was occurring in marctoxctransformation's setup.  Before the patch, the service was never able to find previously held bibs, because although they were being stored properly using the org_code, they were not being read (sql select) properly using the org_code.  (PLEASE NOTE: this scenario only occurs when the tomcat server is restarted after the held holdings record is harvested/processed __and before__ the held bib is harvested/processed.) Now, using the patch, the previously held bibs are found and the necessary holdings record are created.  Also, the mysql table bibsYet2ArriveLongId is now correctly updated by removing the processed (previously held) bib.  Before the patch, this mysql table kept was left untouched (which is "correct" since it wasn't able to process the bib it was supposed to).

There is still one more related bug in this issue.  The SOLR index still tags the holding record as "held" while it should change this value to "active."  So, even though the previously held bibs are now processed correctly, the UI still "thinks" there are "held" bibs even though they really aren't.  

Chris 

Patch submitted as rev. 2855

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-12-22 05:04 PM

Comment body:

John, do you have any input regarding the Solr index still showing records as held.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-12-23 01:13 PM

Comment body:

Ben and I had noticed something (and made a code change) where the status wasn't getting updated properly from Held to Active.  This was for the database though.  I can't find that change, I have been browsing source, looking for diffs, but can't come up with it.  Are you saying that the state of the record is correct in the database, but indexed wrong?

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2011-12-23 01:52 PM

Comment body:

I hadn't looked at the mysql database.  I just noticed that the facet is wrong, which led me to believe that the SOLR index was wrong (since that's what's used for faceting).  I did verify that the records in the SOLR index were wrong...


EDIT:

I take that back somewhat.  I did look at the mysql database, the bibsYet2ArriveLongId table.  Before my patch, the held bibs remained in the table (and they indeed were being held).  After my patch, the table was cleared and the bibs and children were processed correctly, but the SOLR facet kept saying they were held...

Chris


More info:

It's not just a SOLR indexing issue.  The records database for marctoxctransformation is incorrect too:

Before adding the parent bib (notice record id 9 is Held):

mysql> select * from records;
+-----------+---------------------+------+--------+-------------+-----------+
| record_id | oai_datestamp       | type | status | prev_status | format_id |
+-----------+---------------------+------+--------+-------------+-----------+
|         6 | 2011-12-29 09:53:58 | w    | A      | A           |         3 | 
|         7 | 2011-12-29 09:53:58 | e    | A      | A           |         3 | 
|         5 | 2011-12-29 09:53:58 | m    | A      | A           |         3 | 
|         9 | 2011-12-29 09:53:58 | h    | H      | H           |         3 | 
+-----------+---------------------+------+--------+-------------+-----------+
4 rows in set (0.00 sec)

Afterwards (notice record id 9 is still Held, but should've been changed to Active):

mysql> select * from records;
+-----------+---------------------+------+--------+-------------+-----------+
| record_id | oai_datestamp       | type | status | prev_status | format_id |
+-----------+---------------------+------+--------+-------------+-----------+
|         6 | 2011-12-29 09:53:58 | w    | A      | A           |         3 | 
|         7 | 2011-12-29 09:53:58 | e    | A      | A           |         3 | 
|         5 | 2011-12-29 09:53:58 | m    | A      | A           |         3 | 
|         9 | 2011-12-29 09:53:58 | h    | H      | H           |         3 | 
|      1003 | 2012-01-03 10:48:11 | w    | A      | N           |         3 | 
|      1004 | 2012-01-03 10:48:11 | e    | A      | N           |         3 | 
|         8 | 2012-01-03 10:48:11 | m    | A      | N           |         3 | 
+-----------+---------------------+------+--------+-------------+-----------+
7 rows in set (0.00 sec)

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-12-23 02:03 PM

Comment body:

It is something to check, as the indexing code is pretty straightforward, didn't see anything obvious with it, but maybe the change we made to fix this issue didn't get put back properly? 

I had some old notes on this, here they are:

# theoretically, our 124 files that were activated were previously held, we had a bug in our code.
mysql -u root --password=root -D xc_marctoxctransformation -e "select count(*),prev_status from records where status='A' group by prev_status" | less

# wanted to see all updates AFTER t0, looking for our 124 files, try to extract some of the ID's.
mysql -u root --password=root -D xc_marctoxctransformation -e "select count(*), date_updated from record_updates where date_updated > '2011-07-07T12:00:00Z' group by date_updated" | less

# the command run by RepositoryDAO to determine record count totals:
mysql -u root --password=root -D xc_marctoxctransformation -e "select rs.set_id type, r.status status, count(*) c from record_sets rs, records
 r where r.record_id = rs.record_id group by rs.set_id, r.status"

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-03 10:32 PM

Comment body:

I've run through many scenarios and I am now beginning the think the "SOLR indexing" issue was a complete red herring.  It was just a side effect of the mysql database getting into an incorrect state.

Here's a scenario of such incorrect state:


mysql> select * from records;
+-----------+---------------------+------+--------+-------------+-----------+
| record_id | oai_datestamp       | type | status | prev_status | format_id |
+-----------+---------------------+------+--------+-------------+-----------+
|         6 | 2011-12-29 09:53:58 | w    | A      | A           |         3 | 
|         7 | 2011-12-29 09:53:58 | e    | A      | A           |         3 | 
|         5 | 2011-12-29 09:53:58 | m    | A      | A           |         3 | 
|         9 | 2011-12-29 09:53:58 | h    | H      | H           |         3 | 
+-----------+---------------------+------+--------+-------------+-----------+
4 rows in set (0.00 sec)

mysql> select * from held_holdings;
+----------+-----------------+------------------+
| org_code | held_holding_id | manifestation_id |
+----------+-----------------+------------------+
|          |               9 |                8 | 
|          |               9 |                5 | 
+----------+-----------------+------------------+
2 rows in set (0.00 sec)


Notice there are two references in the held_holdings table, one for manifestation_id 8 and one for manifestation_id 5.  But we already processed manifestation_id 5 (see the records table above and you see record_id 5 with type 'm' for manifestation). This is also reflected (correctly) in the bibsYet2Arrive table:

mysql> select * from bibsYet2ArriveLongId;
+----------+---------+-----------+
| org_code | bib_001 | record_id |
+----------+---------+-----------+
| UIUdb    | 1753208 |         8 | 
+----------+---------+-----------+
1 row in set (0.00 sec)

So, when I would later add the bibYet2Arrive (record_id 8) to the repo and have marctoxctransformation process it, it did the correct thing and removed the reference to record_id 8.  And afterward the held_heldings table would shrink by one as well (removing held_holding_id 9 and manifestation_id 8) leaving manifestation id 5, which as I've pointed out at the beginning was already processed and should never have been added to the table to begin with:

mysql> select * from held_holdings;
+----------+-----------------+------------------+
| org_code | held_holding_id | manifestation_id |
+----------+-----------------+------------------+
|          |               9 |                5 | 
+----------+-----------------+------------------+

So, the question really comes down to "how did the mysql table get into this state?"

Sadly, I do not recall.  It's been a long break, etc., etc.

In the meantime, I will try to replicate the scenario above.  And if I can not, then I may have to put this issue on hold (no pun intended) until it resurfaces.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-04 03:30 PM

Comment body:

Chris, I am confused by your final sentence of "until the issue resurfaces."   Doesn't this issue surface all the time?  You have it, we have it, right?  Or did you fix most of this issue and only an oddball case exists that is not reproducible?   If it is reproducible, then it seems it should be fixable.

John, any input?

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-01-04 03:43 PM

Comment body:

Yes I agree this should be readily reproducible.  The change I talked about above (but couldn't find) is in RepositoryDAO, and is in the method I think you should look at:

    public void activateRecords(String name, final TLongHashSet recordIds) {}

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-04 03:53 PM

Comment body:

Randy,

Yes, I'm confused too ;-)

I can't reproduce it.  I'm thinking that it's possible the previous bug fix (in this issue) corrected this problem.  I may have generated the scenario pre-bug fix and can no longer do so (I've been re-using the same VMWare snapshot).

I will now try to undo the bugfix and see if I can reproduce the issue.

Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-04 03:54 PM

Comment body:

John,

I've already looked at (many a times) at that method among others.

This issue allowed me to get to know the code a lot better (and the mysql table relationships).  

Thanks,
Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-04 04:51 PM

Comment body:

I was able to reproduce the scenario.  I had to revert back to code before the bugfix in order to do so.  AFAIK, this may have been the only real bug in this issue.  I'm going to assume this issue is solved until/unless it resurfaces in the future.  Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-01-04 05:53 PM

Comment body:

Maybe we should run this against the UR records and see what our 10,884 held holdings number gets reduced to?   This is probably something I or Chris could kick off on a Friday to run over a weekend.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-11 09:48 PM

Comment body:

I re-ran URs data and only achieved slightly better results (10429 held records).  And of those held records at least one listed in Randy's examples is still incorrectly being held.  I'm re-opening this issue and will now re-focus on the issue using URs dataset.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-01-17 07:41 PM

Comment body:

Fix applied to rev. 2884

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-01-18 04:07 PM

Comment body:

going into 1.2.1

**Comment by patrickzurek** [comment]
Issue resolved: 2012-01-18 04:07 PM
